File: agents/backend/onyx/server/features/character_clothing_changer_ai/static/js/core/rate-limiter.js
/**
 * Rate Limiter Module
 * ==================
 * Rate limiting for API calls and operations
 */

const RateLimiter = {
    /**
     * Rate limiters
     */
    limiters: new Map(),
    
    /**
     * Default options
     */
    defaultOptions: {
        maxRequests: 10,
        windowMs: 60000, // 1 minute
        strategy: 'sliding' // 'sliding' or 'fixed'
    },
    
    /**
     * Create rate limiter
     */
    create(name, options = {}) {
        const config = {
            ...this.defaultOptions,
            ...options
        };
        
        const limiter = {
            name,
            maxRequests: config.maxRequests,
            windowMs: config.windowMs,
            strategy: config.strategy,
            requests: [],
            blocked: false,
            blockUntil: null
        };
        
        this.limiters.set(name, limiter);
        
        if (typeof Logger !== 'undefined') {
            Logger.info(`Rate limiter created: ${name}`, config);
        }
        
        return limiter;
    },
    
    /**
     * Check if request is allowed
     */
    isAllowed(name) {
        const limiter = this.limiters.get(name);
        if (!limiter) {
            return true; // No limiter, allow
        }
        
        const now = Date.now();
        
        // Check if blocked
        if (limiter.blocked && limiter.blockUntil && now < limiter.blockUntil) {
            return false;
        }
        
        // Reset block if expired
        if (limiter.blocked && limiter.blockUntil && now >= limiter.blockUntil) {
            limiter.blocked = false;
            limiter.blockUntil = null;
        }
        
        // Clean old requests based on strategy
        this.cleanRequests(limiter, now);
        
        // Check if limit exceeded
        if (limiter.requests.length >= limiter.maxRequests) {
            // Block for window duration
            limiter.blocked = true;
            limiter.blockUntil = now + limiter.windowMs;
            
            if (typeof Logger !== 'undefined') {
                Logger.warn(`Rate limit exceeded: ${name}`);
            }
            
            // Emit rate limit exceeded event
            if (typeof EventBus !== 'undefined') {
                EventBus.emit('rate-limit:exceeded', { name, limiter });
            }
            
            return false;
        }
        
        // Add request
        limiter.requests.push(now);
        
        return true;
    },
    
    /**
     * Clean old requests
     */
    cleanRequests(limiter, now) {
        if (limiter.strategy === 'sliding') {
            // Remove requests outside window
            limiter.requests = limiter.requests.filter(
                timestamp => now - timestamp < limiter.windowMs
            );
        } else {
            // Fixed window: remove all if window expired
            if (limiter.requests.length > 0) {
                const oldest = limiter.requests[0];
                if (now - oldest >= limiter.windowMs) {
                    limiter.requests = [];
                }
            }
        }
    },
    
    /**
     * Get remaining requests
     */
    getRemaining(name) {
        const limiter = this.limiters.get(name);
        if (!limiter) {
            return Infinity;
        }
        
        const now = Date.now();
        this.cleanRequests(limiter, now);
        
        return Math.max(0, limiter.maxRequests - limiter.requests.length);
    },
    
    /**
     * Get reset time
     */
    getResetTime(name) {
        const limiter = this.limiters.get(name);
        if (!limiter) {
            return null;
        }
        
        if (limiter.requests.length === 0) {
            return null;
        }
        
        const oldest = limiter.requests[0];
        return oldest + limiter.windowMs;
    },
    
    /**
     * Reset limiter
     */
    reset(name) {
        const limiter = this.limiters.get(name);
        if (!limiter) {
            return false;
        }
        
        limiter.requests = [];
        limiter.blocked = false;
        limiter.blockUntil = null;
        
        if (typeof Logger !== 'undefined') {
            Logger.info(`Rate limiter reset: ${name}`);
        }
        
        return true;
    },
    
    /**
     * Get limiter status
     */
    getStatus(name) {
        const limiter = this.limiters.get(name);
        if (!limiter) {
            return null;
        }
        
        const now = Date.now();
        this.cleanRequests(limiter, now);
        
        return {
            name: limiter.name,
            maxRequests: limiter.maxRequests,
            currentRequests: limiter.requests.length,
            remaining: limiter.maxRequests - limiter.requests.length,
            blocked: limiter.blocked,
            blockUntil: limiter.blockUntil,
            resetTime: this.getResetTime(name)
        };
    },
    
    /**
     * Get all limiters
     */
    getAll() {
        return Array.from(this.limiters.values());
    }
};

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = RateLimiter;
}

File: autonomous_long_term_agent/core/rate_limiter.py
"""
Rate Limiting for API endpoints
"""

import asyncio
import time
import logging
from collections import defaultdict
from typing import Dict, Tuple
from datetime import datetime, timedelta

logger = logging.getLogger(__name__)


class RateLimiter:
    """Simple rate limiter using token bucket algorithm"""
    
    def __init__(
        self,
        max_requests: int = 100,
        window_seconds: int = 60
    ):
        self.max_requests = max_requests
        self.window_seconds = window_seconds
        self._requests: Dict[str, list] = defaultdict(list)
        self._lock = asyncio.Lock()
    
    async def is_allowed(self, key: str = "default") -> Tuple[bool, int]:
        """
        Check if request is allowed
        
        Returns:
            Tuple of (is_allowed, remaining_requests)
        """
        async with self._lock:
            now = time.time()
            cutoff = now - self.window_seconds
            
            # Clean old requests
            self._requests[key] = [
                req_time for req_time in self._requests[key]
                if req_time > cutoff
            ]
            
            # Check if limit exceeded
            if len(self._requests[key]) >= self.max_requests:
                remaining = 0
                return False, remaining
            
            # Add current request
            self._requests[key].append(now)
            remaining = self.max_requests - len(self._requests[key])
            
            return True, remaining
    
    async def get_stats(self, key: str = "default") -> Dict[str, any]:
        """Get rate limit statistics"""
        async with self._lock:
            now = time.time()
            cutoff = now - self.window_seconds
            
            requests = [
                req_time for req_time in self._requests[key]
                if req_time > cutoff
            ]
            
            return {
                "requests_in_window": len(requests),
                "max_requests": self.max_requests,
                "window_seconds": self.window_seconds,
                "remaining": max(0, self.max_requests - len(requests))
            }




